File: neetcode150/15_greedy/0045_jump_game_ii.cpp
/*
 * =====================================================================
 * LeetCode 45 : Jump Game II                                        (Medium)
 * https://leetcode.com/problems/jump-game-ii/
 * Category   : Greedy
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Start at index 0 of `nums`; from i you can jump at most nums[i]
 *   steps forward. Return the MINIMUM number of jumps to reach the
 *   last index. It is guaranteed the end is reachable.
 * ---------------------------------------------------------------------
 * INTUITION
 *   Think in LAYERS (BFS on the implicit jump graph). Layer k = all
 *   positions reachable with exactly ≤ k jumps. The frontier of layer
 *   k is an interval [l, r]; one more jump from ANY cell inside it
 *   covers at most max(i + nums[i]) over the interval — the greedy
 *   jump is the one that maximizes that single value. Taking the
 *   furthest option per layer is optimal because the next layer's
 *   frontier is exactly (r, maxReach]: a shorter jump only lands
 *   inside [l, r] again (no progress) or covers less ground (worse
 *   frontier) — exchange argument.
 * ---------------------------------------------------------------------
 * APPROACH 1 — BFS-layered greedy (BEST, O(1) space)
 *     curEnd = right edge of the CURRENT layer; far = furthest any
 *     cell in this layer can reach. When i reaches curEnd the layer
 *     is exhausted → one jump consumed, curEnd = far.
 *   Time  Complexity : O(n)   Space Complexity : O(1)
 *   WHY correctness: layer k is an interval. Minimal jumps = number
 *   of layers until the interval contains the end — greedy maximizes
 *   each layer's right edge, so it minimizes the layer count.
 * ---------------------------------------------------------------------
 * APPROACH 2 — Reverse greedy O(n²) (pick leftmost reachable)
 *   Work from the end backwards: repeatedly find the LEFTMOST index
 *   that can jump to the current target (greedy-leftmost is correct
 *   because landing further left can only leave more options open).
 *   Time  Complexity : O(n²)   Space Complexity : O(1)
 *   Correct but quadratic: the leftmost-reachable search re-scans the
 *   array once per jump. Forward BFS-layered does the same counting
 *   in a single sweep.
 * ---------------------------------------------------------------------
 * DEEP DIVE — why per-layer furthest is the right greed
 *   Exchange argument: in layer k, every cell c in [l, r] is ≤ r, and
 *   its reach c + nums[c] ≤ maxReach (the greedy pick). So the greedy
 *   next-layer frontier [r, maxReach] CONTAINS the frontier any other
 *   choice would produce → any path the rival continues from, we can
 *   continue from too. Greedy layer count ≤ rival layer count, and
 *   since both are valid (≥ optimal), greedy is optimal. This is the
 *   same interval-frontier monotonicity as Jump Game I, upgraded to
 *   COUNTING layers.
 * =====================================================================
 */

#include <iostream>       // std::cout — benchmark output
#include <vector>         // std::vector — input array
#include <string>         // std::string — PASS/FAIL messages
#include <chrono>         // std::chrono — high-resolution timing
#include <sys/resource.h> // getrusage — peak RSS measurement

/* =====================================================================
 * APPROACH 1 : BFS-layered greedy (furthest reach per layer)
 * ===================================================================== */
class Solution_1 {
public:
    /*
     * Purpose : Return the minimum number of jumps to reach the end.
     * Inputs  : nums — vector of ints, nums[i] = max forward jump.
     * Output  : int — minimal jump count (end guaranteed reachable).
     */
    int jump(const std::vector<int>& nums) {
        int jumps = 0;  // layers consumed so far
        int curEnd = 0; // right edge of the CURRENT jump layer
        int far = 0;    // furthest cell the current layer can reach

        // We never need to process the last cell: being there means
        // we are already done (its jump value is irrelevant).
        for (int i = 0; i < static_cast<int>(nums.size()) - 1; ++i) {
            if (i + nums[i] > far) far = i + nums[i]; // widen next frontier

            if (i == curEnd) {
                ++jumps;      // current layer exhausted → must take a jump
                curEnd = far; // its frontier becomes the next layer
            }
        }
        return jumps;
    }
};

/* =====================================================================
 * APPROACH 2 : Reverse greedy — hop backwards to leftmost reachable
 * ===================================================================== */
class Solution_2 {
public:
    /*
     * Purpose : Return the minimum number of jumps to reach the end.
     * Inputs  : nums — vector of ints, nums[i] = max forward jump.
     * Output  : int — minimal jump count (end guaranteed reachable).
     */
    int jump(const std::vector<int>& nums) {
        int jumps = 0;
        int pos = static_cast<int>(nums.size()) - 1; // target to reach

        // Each iteration places `pos` one jump closer to index 0.
        while (pos > 0) {
            // Leftmost index that can jump ONTO pos. Landing leftmost
            // is safest: it leaves the most future options open.
            for (int i = 0; i < pos; ++i) {
                if (i + nums[i] >= pos) {
                    pos = i;    // new target = the takeoff cell
                    ++jumps;
                    break;      // restart search from the new position
                }
            }
        }
        return jumps;
    }
};

/* =====================================================================
 * BENCHMARK — run both approaches on 3 tests, time + peak RAM
 * ===================================================================== */
int main() {
    // Official examples + a pathological all-ones array (n-1 jumps).
    const std::vector<std::vector<int> > tests = {
        { 2, 3, 1, 1, 4 }, // → 2  (0→1→4)
        { 2, 3, 0, 1, 4 }, // → 2  (0→1→4)
        { 1, 1, 1, 1 },    // → 3  (walk one step each time)
        { 0 },             // → 0  (already at the end)
        { 1, 2, 3 }        // → 2  (0→1→2; direct 0→2 impossible, jump=1)
    };
    const std::vector<int> expected = { 2, 2, 3, 0, 2 };

    auto run_all = [&](int which, int (*solve)(const std::vector<int>&)) {
        struct rusage before, after;
        getrusage(RUSAGE_SELF, &before);
        auto t0 = std::chrono::high_resolution_clock::now();

        bool ok = true;
        for (int t = 0; t < static_cast<int>(tests.size()); ++t) {
            if (solve(tests[t]) != expected[t]) ok = false;
        }

        auto t1 = std::chrono::high_resolution_clock::now();
        getrusage(RUSAGE_SELF, &after);
        double us = std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count() / 1000.0;
        long mem_kb = after.ru_maxrss - before.ru_maxrss; // delta, KB

        std::cout << "Approach " << which
                  << " : " << (ok ? "PASS" : "FAIL")
                  << "  time = " << us << " µs  peak-ram-delta = " << mem_kb << " KB\n";
        return ok;
    };

    bool a1 = run_all(1, [](const std::vector<int>& v) { return Solution_1().jump(v); });
    bool a2 = run_all(2, [](const std::vector<int>& v) { return Solution_2().jump(v); });
    std::cout << (a1 && a2 ? "PASS : both approaches return the minimal jump count."
                           : "FAIL : at least one approach is wrong.") << "\n";
    return 0;
}

File: neetcode150/15_greedy/0045_jump_game_ii.py
"""
=====================================================================
LeetCode 45 : Jump Game II                                     (Medium)
https://leetcode.com/problems/jump-game-ii/
Category   : Greedy
---------------------------------------------------------------------
PROBLEM
    Start at index 0 of `nums`; from i you can jump at most nums[i]
    steps forward. Return the MINIMUM number of jumps to reach the
    last index. The end is guaranteed reachable.
---------------------------------------------------------------------
INTUITION
    Think in BFS LAYERS over the implicit jump graph. Layer k = all
    positions reachable with ≤ k jumps — always one contiguous
    interval [l, r]. One more jump from any cell in the interval
    reaches at most max(i + nums[i]); the greedy jump maximizes that
    single frontier value, and maximizing every layer's right edge
    minimizes the number of layers needed → minimal jumps.
---------------------------------------------------------------------
APPROACH 1 — BFS-layered greedy (BEST)
    curEnd = current layer's right edge; far = furthest any cell in
    the layer can reach. When i reaches curEnd, one jump is consumed
    and curEnd = far.
    Time  Complexity : O(n)   Space Complexity : O(1)
---------------------------------------------------------------------
APPROACH 2 — Reverse greedy (leftmost reachable, O(n²))
    Work backwards: repeatedly find the LEFTMOST index that can jump
    onto the current target. Landing leftmost leaves the most future
    options open → provably optimal, but re-scans per jump.
    Time  Complexity : O(n²)   Space Complexity : O(1)
---------------------------------------------------------------------
DEEP DIVE — why furthest-per-layer is provably optimal
    Exchange argument: every cell c in layer k satisfies c ≤ r, so
    c + nums[c] ≤ maxReach. The greedy frontier [r, maxReach] CONTAINS
    the frontier any other choice would create → any path a rival
    follows, we can follow too → greedy needs no more jumps than any
    rival. Interval-frontier monotonicity (same as Jump Game I) is the
    mathematical backbone.
---------------------------------------------------------------------
REAL LIFE
    • Router/CDN hop planning: minimum hops to deliver a packet.
    • EV range planning: minimal charging stops given per-stop range.
    • Game speed-run routing: minimal moves to a goal.
=====================================================================
"""

from __future__ import annotations
import time
import tracemalloc
from typing import List

# =====================================================================
# APPROACH 1 : BFS-layered greedy (best)
# =====================================================================
class SolutionOne:
    """
    Purpose : Return the minimum number of jumps to reach the end.
    Inputs  : nums — list of ints, nums[i] = max forward jump length.
    Output  : int — minimal jump count (end guaranteed reachable).
    """

    def jump(self, nums: List[int]) -> int:
        jumps = 0   # layers (jumps) consumed so far
        cur_end = 0 # right edge of the current jump layer
        far = 0     # furthest cell the current layer can reach

        # Last cell never needs processing — arriving there ends it.
        for i in range(len(nums) - 1):
            far = max(far, i + nums[i])  # widen next frontier from here

            if i == cur_end:             # current layer fully swept
                jumps += 1               # must take exactly one more jump
                cur_end = far            # its frontier = next layer
        return jumps

# =====================================================================
# APPROACH 2 : Reverse greedy — hop backwards to leftmost reachable
# =====================================================================
class SolutionTwo:
    """
    Purpose : Return the minimum number of jumps to reach the end.
    Inputs  : nums — list of ints, nums[i] = max forward jump length.
    Output  : int — minimal jump count (end guaranteed reachable).
    """

    def jump(self, nums: List[int]) -> int:
        jumps = 0
        pos = len(nums) - 1  # target cell to reach from the left

        # Each pass lands `pos` one jump closer to index 0.
        while pos > 0:
            # Leftmost takeoff cell that can jump onto pos — landing
            # leftmost keeps the most future options open.
            for i in range(pos):
                if i + nums[i] >= pos:
                    pos = i
                    jumps += 1
                    break   # restart the search from the new target
        return jumps


# =====================================================================
# BENCHMARK — both approaches, time + peak memory per approach
# =====================================================================
if __name__ == "__main__":
    # Official examples + all-ones walk + already-at-end + direct hop.
    tests: List[List[int]] = [
        [2, 3, 1, 1, 4],  # → 2
        [2, 3, 0, 1, 4],  # → 2
        [1, 1, 1, 1],     # → 3
        [0],              # → 0
        [1, 2, 3],        # → 2  (0→1→2; jump=1 forbids direct hop)
    ]
    expected = [2, 2, 3, 0, 2]

    def run_all(solve) -> tuple[bool, float, int]:
        """Time one solver over every test; return (allPass, µs, peakBytes)."""
        tracemalloc.start()
        t0 = time.perf_counter()
        ok = all(solve(t) == e for t, e in zip(tests, expected))
        t1 = time.perf_counter()
        _, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        return ok, (t1 - t0) * 1e6, peak

    ok1, us1, peak1 = run_all(lambda t: SolutionOne().jump(t))
    ok2, us2, peak2 = run_all(lambda t: SolutionTwo().jump(t))

    print(f"Approach 1 (BFS-layered)   : {'PASS' if ok1 else 'FAIL'}  time = {us1:.2f} µs  peak-mem = {peak1} bytes")
    print(f"Approach 2 (reverse greedy) : {'PASS' if ok2 else 'FAIL'}  time = {us2:.2f} µs  peak-mem = {peak2} bytes")
    print("PASS : both approaches return the minimal jump count."
          if ok1 and ok2
          else "FAIL : at least one approach is wrong.")
